blog/_posts/2017-02-25-collections-rework.md

```diff
@@ -0,0 +1,219 @@
+
+Views are also the recommended replacement for `collection.breakOut`. For example,
+
+```scala
+val s: Seq[Int] = ...
+val set: Set[String] = s.map(_.toString)(collection.breakOut)
+```
+
+can be expressed with the same performance characteristics as:
+
+```scala
+val s: Seq[Int] = ...
+val set = s.view.map(_.toString).to(Set)
+```
+
+If you combine multiple operations they are executed lazily:
+
+```scala
+val s: Seq[Int] = ...
+
+// First build a filtered Seq, then a mapped Seq, then a Set
+val set1 = s.filter(_ > 0).map(_.toString).to(Set)
+
+// Build a Set directly by evaluating filter and map together:
+val set2 = s.view.filter(_ > 0).map(_.toString).to(Set)
+```
+
+## Laziness
+
+Aside from views the current collections library also supports lazy collections but they do not get a first class treatment. While we do have `Stream`, it is only lazy in the tail but still strict in the head element and the implementation needs to special-case pretty much everything because the basic abstraction for building new collections (which is used by all of the default implementations that strict collection types can safely inherit) is `CanBuildFrom` / `Builder` which uses strict, push-based collection building.
+
+By combining `FromIterable` with the new `View` design we can turn this around to provide pull-based building which can be used by strict and lazy collections alike.  For example, this is the default implementation of `map` in `IterablePolyTransforms` which does not need to be overridden in `LazyList` to provide the expected laziness:
+
+```scala
+  def map[B](f: A => B): C[B] = fromIterable(View.Map(coll, f))
+```
+
+Any operation that is available on `View` can get a default implementation like this based on rebuilding the current collection type from a `View` operation. Naturally a `LazyList` can be "built" from an `Iterable` by getting an `Iterator` and only pulling elements one by one as they are needed, so we automatically get a lazy implementation of `map`.
+
+## Language Integration
+
+You may have wondered about the asymmetry in the default types that are in scope through `Predef` or the `scala` package object:
+
+```
+scala> classOf[Set[_]]
+res1: Class[Set[_]] = interface scala.collection.immutable.Set
+
+scala> classOf[Map[_, _]]
+res2: Class[Map[_, _]] = interface scala.collection.immutable.Map
+
+scala> classOf[Seq[_]]
+res3: Class[Seq[_]] = interface scala.collection.Seq
+```
+
+Unlike all other collection types, `Seq` is not the immutable version but the generic one that encompasses both, mutable and immutable sequences. The reason for this is that the Scala specification represents varargs as type `scala.Seq` (and it should not have to rely on any type outside the top-level `scala` package). Since varargs in Java (and on the JVM) are really mutable arrays, the default `Seq` has to allow mutable collections.
```

Comment:
Yes, that's correct, `scala.Seq` == `scala.collection.Seq`, unlike `Predef.Set` and `Predef.Map` which are aliases to `scala.collection.immutable.Set` and `scala.collection.immutble.Map` respectively.